psa-crypto/src/types/key_derivation_operation.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! # PSA Operation types
+
+use super::algorithm::{Hash, KeyDerivation};
+use super::key::Id;
+#[cfg(feature = "interface")]
+use super::status::{Error, Result, Status};
+#[cfg(feature = "interface")]
+use core::convert::{From, TryFrom};
+
+/// Key derivation operation for deriving keys from existing sources
+#[derive(Debug, Clone)]
+pub struct KeyDerivationOperation<'a> {
+    /// Key derivation algorithm and inputs
+    pub alg: KeyDerivationWithInputs<'a>,
+    /// Maximum capacity of a key derivation operation
+    pub capacity: Option<usize>,
+}
+
+/// Wrapper around KeyDerivation to enforce correct `Input`s
+#[derive(Debug, Clone)]
+pub enum KeyDerivationWithInputs<'a> {
+    /// HKDF algorithm.
+    Hkdf {
+        /// A hash algorithm to use
+        hash_alg: Hash,
+        /// Salt, used in the "extract" step. It is optional; if omitted, the derivation uses an empty salt.
+        /// Typically a direct input, can also be a key of type `RawData`.
+        salt: Option<Input<'a>>,
+        /// Secret, used in the "extract" step. Must be `Input::Key(Id)` if used with `psa_key_derivation_output_key`.
+        secret: Input<'a>,
+        /// Info, used in the "expand" step. Typically a direct input, can also be a key of type `RawData`.
+        info: Input<'a>,
+    },
+    /// TLS-1.2 PRF algorithm.
+    Tls12Prf {
+        /// A hash algorithm to use.
+        hash_alg: Hash,
+        /// Seed, typically a direct input, can also be a key of type `RawData`.
+        seed: Input<'a>,
+        /// Secret. Must be `Input::Key(Id)` if used with `psa_key_derivation_output_key`.
+        secret: Input<'a>,
+        /// Label. Typically a direct input, can also be a key of type `RawData`.
+        label: Input<'a>,
+    },
+    /// TLS-1.2 PSK-to-MasterSecret algorithm.
+    Tls12PskToMs {
+        /// A hash algorithm to use.
+        hash_alg: Hash,
+        /// Seed, typically a direct input, can also be a key of type `RawData`.
+        seed: Input<'a>,
+        /// Secret. Must be `Input::Key(Id)` if used with `psa_key_derivation_output_key`.
+        /// Must not be larger than `PSA_TLS12_PSK_TO_MS_PSK_MAX_SIZE`.
+        secret: Input<'a>,
+        /// Label. Typically a direct input, can also be a key of type `RawData`.
+        label: Input<'a>,
+    },
+}
+
+/// Enumeration of the step of a key derivation.
+#[derive(Debug, Clone, Copy)]
+pub enum DerivationStep {
```

Comment:
This could be `Step` 🧹

Reply:
I don't know about this one - not sure if `Step` is so common a term in this kind of context to be fully clear what it means without the `Derivation` part of the name